### PR TITLE
c-parser+crefine: move h_t_array_valid_field to c-parser session

### PR DESCRIPTION
**Pull request by michaelmcinerney**

### CODE REVIEW

**Review by Xaphiosis — APPROVED** [review]
Looks good.

File: tools/c-parser/umm_heap/ArrayAssertion.thy [hunk 1]
```diff
@@ -172,4 +172,54 @@ lemma ptr_arr_retyps_to_retyp:
     \<Longrightarrow> ptr_arr_retyps n (p :: ('c :: wf_type) ptr) = ptr_retyp (ptr_coerce p :: ('c['b]) ptr)"
   by (auto simp: ptr_arr_retyps_def ptr_retyp_def typ_slices_def typ_uinfo_array_tag_n_m_eq)
 
+lemma size_td_uinfo_array_tag_n_m[simp]:
+  "size_td (uinfo_array_tag_n_m (ta :: ('a :: c_type) itself) n m)
```
Comment (lsf37) [inline tools/c-parser/umm_heap/ArrayAssertion.thy:176]:
```suggestion
lemma size_td_uinfo_array_tag_n_m[simp]:
  "size_td (uinfo_array_tag_n_m (t :: 'a :: c_type itself) n m)
```

File: tools/c-parser/umm_heap/ArrayAssertion.thy [hunk 2]
```diff
@@ -172,4 +172,54 @@ lemma ptr_arr_retyps_to_retyp:
...
+     apply auto[1]
+    apply (simp add: o_def)
+    apply (subst Suc.hyps)
+     apply arith
+    apply (metis mod_geq)
+    done
+qed
+
+lemma h_t_array_valid_field:
+  "h_t_array_valid htd (p :: ('a :: wf_type) ptr) n
```
Comment (lsf37) [inline tools/c-parser/umm_heap/ArrayAssertion.thy:205]:
```suggestion
lemma h_t_array_valid_field:
  "h_t_array_valid htd (p :: 'a :: wf_type ptr) n
```

**Review by lsf37 — APPROVED** [review]
Looks good, just added a bit of cleanup that we should do if we touch these.

**Pull request merged**